File: 2003/VB101SamplesAll/VB101SamplesBCL2/CreateAProcessViewer/How-To Process Viewer/frmModules.vb
'Copyright (C) 2002 Microsoft Corporation
'All rights reserved.
'THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER 
'EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED WARRANTIES OF 
'MERCHANTIBILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.

'Requires the Trial or Release version of Visual Studio 2005.


Public Class frmModules
    Private m_ParentProcess As Process = Nothing

    Private mits As ListView.ListViewItemCollection
    Private mcolModules As New Collection()

    Friend Property ParentProcess() As Process
        Get
            Return m_ParentProcess
        End Get
        Set(ByVal Value As Process)
            m_ParentProcess = Value
            If m_ParentProcess Is Nothing Then
                mcolModules = Nothing
            End If
        End Set
    End Property

    Private Sub EnumModules()
        Try
            Me.lvModules.Items.Clear()
            If Not mcolModules Is Nothing Then
                mcolModules = New Collection()
            End If
            Dim m As ProcessModule
            For Each m In m_ParentProcess.Modules
                Me.lvModules.Items.Add(m.ModuleName)

                Try
                    mcolModules.Add(m, m.ModuleName)
                Catch exp As ArgumentException
                    ' This means the item was duplicated.
                    ' Eat error and continue.
                Catch exp As Exception
                    MessageBox.Show(exp.Message, exp.Source, MessageBoxButtons.OK, MessageBoxIcon.Error)
                End Try

            Next
        Catch exp As Exception
            MessageBox.Show(exp.Message, exp.Source, MessageBoxButtons.OK, MessageBoxIcon.Error)
        End Try
    End Sub

    Friend Sub RefreshModules()
        Me.sbInfo.Text = "Process = " & m_ParentProcess.ProcessName
        Me.lvModDetail.Items.Clear()
        EnumModules()
    End Sub

    Private Sub mnuClose_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles mnuClose.Click
        Me.Hide()
    End Sub

    Private Sub EnumModule(ByVal m As ProcessModule)

        Me.lvModDetail.Items.Clear()

        mits = Me.lvModDetail.Items

        Try
            AddNameValuePair("Base Address", Hex(m.BaseAddress.ToInt32).ToLower())
            AddNameValuePair("Entry Point Address", Hex(m.EntryPointAddress.ToInt32).ToLower())
            AddNameValuePair("File Name", m.FileName)
            AddNameValuePair("File Version", m.FileVersionInfo.FileVersion.ToString())
            AddNameValuePair("File Description", m.FileVersionInfo.FileDescription)
            AddNameValuePair("Memory Size", m.ModuleMemorySize.ToString("N0"))

        Catch exp As Exception
            MessageBox.Show(exp.Message, exp.Source, MessageBoxButtons.OK, MessageBoxIcon.Error)
        End Try
    End Sub

    Private Sub AddNameValuePair(ByVal Item As String, ByVal SubItem As String)
        With mits.Add(Item)
            .SubItems.Add(SubItem)
        End With
    End Sub

    Private Sub lvModules_SelectedIndexChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles lvModules.SelectedIndexChanged, lvModules.Click
        Try
            Dim lv As ListView = CType(sender, ListView)

            If lv.SelectedItems.Count = 1 Then
                Dim strMod As String = lv.SelectedItems(0).Text

                Dim m As ProcessModule = CType(mcolModules.Item(strMod), ProcessModule)
                EnumModule(m)
            End If
        Catch exp As Exception
            MessageBox.Show(exp.Message, exp.Source, MessageBoxButtons.OK, MessageBoxIcon.Error)
        End Try
    End Sub
End Class

File: 2003/CS101SamplesAll/CS101SamplesBCL2/CreateAProcess Viewer/How-To Process Viewer/frmModules.cs
using System;
using System.Windows.Forms;
using System.Diagnostics;
using System.Collections;
using System.Collections.Generic;

namespace How_To_Process_Viewer
{
    public partial class frmModules : Form
    {
        public frmModules()
        {
            InitializeComponent();
        }
        private Process m_ParentProcess;

        private ListView.ListViewItemCollection mits;

        private List<ProcessModule> _mcolModules = new List<ProcessModule>();

        public Process ParentProcess
        {
            get
            {
                return ParentProcess;
            }
            set
            {
                m_ParentProcess = value;

                if (m_ParentProcess == null)
                {
                    _mcolModules = null;
                }

            }

        }

        private void EnumModules()
        {
            try
            {
                this.lvModules.Items.Clear();

                if (!(_mcolModules == null))
                {
                    _mcolModules = new List<ProcessModule>();
                }

                foreach (ProcessModule m in m_ParentProcess.Modules)
                {
                    this.lvModules.Items.Add(m.ModuleName);
                    try
                    {
                        _mcolModules.Add(m);
                    }
                    catch (Exception exp)
                    {
                        MessageBox.Show(exp.Message, exp.Source, MessageBoxButtons.OK, MessageBoxIcon.Error);
                    }
                }
            }
            catch (Exception exp)
            {
                MessageBox.Show(exp.Message, exp.Source, MessageBoxButtons.OK, MessageBoxIcon.Error);
            }
        }

        public void RefreshModules()
        {
            this.sbInfo.Text = "Process = " + m_ParentProcess.ProcessName;
            this.lvModDetail.Items.Clear();
            EnumModules();
        }

        private void mnuClose_Click(object sender, System.EventArgs e)
        {
            this.Hide();
        }

        private void EnumModule(ProcessModule m)
        {
            this.lvModDetail.Items.Clear();
            mits = this.lvModDetail.Items;
            try
            {
                AddNameValuePair("Base Address", m.BaseAddress.ToInt32().ToString("x").ToLower());
                AddNameValuePair("Entry Point Address", m.EntryPointAddress.ToInt32().ToString("x").ToLower());
                AddNameValuePair("File Name", m.FileName);
                AddNameValuePair("File Version", m.FileVersionInfo.FileVersion.ToString());
                AddNameValuePair("File Description", m.FileVersionInfo.FileDescription);
                AddNameValuePair("Memory Size", m.ModuleMemorySize.ToString("N0"));
            }
            catch (Exception exp)
            {
                MessageBox.Show(exp.Message, exp.Source, MessageBoxButtons.OK, MessageBoxIcon.Error);
            }
        }

        private void AddNameValuePair(string Item, string voidItem)
        {
            mits.Add(Item).SubItems.Add(voidItem);
        }

        private void lvModules_SelectedIndexChanged(object sender, System.EventArgs e)
        {
            try
            {
                ListView lv = ((ListView)sender);
                if (lv.SelectedItems.Count == 1)
                {
                    string strMod = lv.SelectedItems[0].Text;
                    int ItemIndex = lv.FocusedItem.Index;
                    ProcessModule m = ((ProcessModule)(_mcolModules[ItemIndex]));
                       
                    EnumModule(m);
                }
            }
            catch (Exception exp)
            {
                MessageBox.Show(exp.Message, exp.Source, MessageBoxButtons.OK, MessageBoxIcon.Error);
            }
        }
    }
}